### ISSUE TITLE
Execute child templates only after the templating engines

### ISSUE THREAD

**Issue by pofider** [jsreport/jsreport#545]
Currently, we evaluate child templates twice.
Once in beforeRenderListeners, so before the templating engines.
Once in afterTemplatingEnginesListeners, so after templating engines.

Maybe there is currently no need to evaluate it twice.
I have a feeling we run evaluation in beforeRender just because to replicate the usage of the assets. And maybe there is no need for it when static assets are already part of jsreport.

Rendering just in afterTemplatingEnginesListeners makes working this use case for example:

```
{{#if foo}}
   {#child template1}
{{else}}
  {#child template2}
{{/if}}
```
Currently, both child templates are evaluated and we don't want that.

**Comment by wazowsk1** [comment]
Also child templates are evaluated when in comments, which I would not expect
e.g.
```
{{!--  {#child template1} --}}
 ```

**Comment by pofider** [comment]
@wazowsk1 Thank you for the input. Yes, this task should solve also this.

**Comment by wazowsk1** [comment]
Given "foo" in your example is false and the child template1 uses a property from the data model
When I do not provide the property for this child template
Then rendering will fail

Is there any roadmap for v3 or this specific feature? due the performance impact and data model requirements. 
I have limited time, but let me know if I can help with this feature.

**Comment by pofider** [comment]
Yes, it will be part of the v3.
At this moment you can comment out this line
https://github.com/jsreport/jsreport-child-templates/blob/master/lib/childTemplates.js#L65

Or I believe you can also create a jsreport custom script, attach it to the child template and skip rendering under some conditions. Like this

```
function beforeRender(req, res) {
   if (req.data.foo === false) {
      req.template.content = ''
   }
}
```

**Comment by pofider** [comment]
We will actually deprecate the `{#child template1}` syntax in v3.  Don't panic, it will be still supported.
Since v3 the child templates will be invoked using standard helpers call `{{childTemplate xxx}}`.
This is possible because we finally support async helpers.
This will solve problems with comments and passing input data.

This means we won't do changes in the old syntax and keep it as it is, in the deprecation mode.
